### PR TITLE
Convert DaysBetween to full template goodie

### PR DESCRIPTION
**Pull request by NaveenKarippai**
###### What does your Pull Request do (check all that apply)?

Choose the most relevant items and use the following title template to name
your Pull Request.

- [ ] New Instant Answer
    - [ ] Cheat Sheets: **`New {Cheat Sheet Name} Cheat Sheet`**
    - [ ] Other: **`New {IA Name} Instant Answer`**
- [x] Improvement
    - [ ] Bug fix: **`{IA Name}: Fix {Issue number or one-line description}`**
    - [x] Enhancement: **`days_between: use full-template, refactor test suite`**
- [ ] Non–Instant Answer
    - [ ] Other (Role, Template, Test, Documentation, etc.): **`{GoodieRole/Templates/Tests/Docs}: {Short Description}`**

###### Description of changes

Goodies have full support for templates, and it'd be great
to make sure all Goodies use full templates.
The "simple structured_answer" was provided as a temporary
workaround.

Provide an overview of the changes this pull request introduces.

* change to 'text' template for structured_answer object

* refactored test suite

###### Which issues (if any) does this fix?

Re: #2767

Fixes #NNNN - how specifically does it fix the issue?

###### People to notify (@GuiltyDolphin)

---

Instant Answer Page: https://duck.co/ia/view/days_between

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @JetFault

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: t/DaysBetween.t [hunk 1]
```diff
@@ -10,143 +10,42 @@ use Test::MockTime qw( :all );
...
-            operation => 'Days between',
-            result    => 53
-        },
-    ),
+    'days between 01/01/2000 01/01/2001'                        => build_test('01 Jan 2000', '01 Jan 2001', 366, '' ),
+    'days between 1/1/2000 and 1/1/2001 inclusive'              => build_test('01 Jan 2000', '01 Jan 2001',367, $test_inclusive),
+    'daysbetween 03/4/2005 and 11/8/2020'                       => build_test('04 Mar 2005', '08 Nov 2020', 5728, ''),
+    'days_between 3/14/2005 and 1/2/2003'                       => build_test('02 Jan 2003', '14 Mar 2005', 802, ''),
+    'days between 01/31/2000 01/31/2001'                        => build_test('31 Jan 2000', '31 Jan 2001', 366, ''),
+    'days between 01/31/2000 01/31/2001 inclusive'              => build_test('31 Jan 2000', '31 Jan 2001', 367, $test_inclusive),
```
Comment (GuiltyDolphin) [inline t/DaysBetween.t]: Lets replace all these ambiguous dates with ISO8601 format (`YYYY-MM-DD`)

Comment (NaveenKarippai) [inline t/DaysBetween.t]: @GuiltyDolphin Thanks for the review! updated to ISO8601 format as suggested.

File: t/DaysBetween.t [hunk 2]
```diff
@@ -10,143 +10,42 @@ use Test::MockTime qw( :all );
...
+    'days between 2000-01-31 2001-01-31'                        => build_test('31 Jan 2000', '31 Jan 2001', 366, ''),
+    'days between 2000-01-31 2001-01-31 inclusive'              => build_test('31 Jan 2000', '31 Jan 2001', 367, $test_inclusive),
+    'days between January 31st, 2000 and 31-Jan-2001 inclusive' => build_test('31 Jan 2000', '31 Jan 2001', 367, $test_inclusive),
+    'days between jan 1 2012 and jan 1 1234'                    => build_test('01 Jan 1234', '01 Jan 2012', 284158, ''),
+    'days between jan 1 and jan 15 inclusive'                   => build_test('01 Jan 2016', '15 Jan 2016', 15, $test_inclusive),
+    'days between jan 1 and 15th feb'                           => build_test('01 Jan 2016', '15 Feb 2016', 45, ''),
+    'days between today and tomorrow'                           => build_test('14 Jul 2016', '15 Jul 2016', 1, ''),
+    'how many days between feb 2 and feb 17'                    => build_test('02 Feb 2016', '17 Feb 2016', 15, ''),
+    'days between 22nd may and today'                           => build_test('22 May 2016', '14 Jul 2016', 53, ''),
+    'days between jan 1 2012 and jan 1 123456'                  => undef
```
Comment (GuiltyDolphin) [inline t/DaysBetween.t]: @NaveenKarippai Hmm, I think we can probably get rid of most of these - we only need to test a few cases:

1. First date is earlier than second.
2. Second date is earlier than first.
3. Dates represent the same day.
4. Inclusive versions of those above.

Lets pull a test for each of those listed, and remove the rest - most of these just seem to be testing the parsing (which is handled externally - by the Role).

/cc @mintsoft

Comment (mintsoft) [inline t/DaysBetween.t]: Also with leapyears (if it's covered by these tests) and we should keep the ambiguous tests:
* `days between today and tomorrow`
* `days between jan 1 and jan 15 inclusive`
* `days between jan 1 and 15th feb`
* `'how many days between feb 2 and feb 17'`

As these are either question formats or form the contract with the role

Comment (GuiltyDolphin) [inline t/DaysBetween.t]: @mintsoft Yeah, I meant to include the leap year tests, whoops!

I think any ability to parse dates proves the Role is being used - unless you are going out of your way to use separate parsing I think just the ISO8601 format should be fine; that being said I guess you *could* be *only* allowing formatted dates through... Gnh, I guess one test using a non-ambiguous, but non 'formatted' date should be fine.

Comment (mintsoft) [inline t/DaysBetween.t]: @GuiltyDolphin Yeah that sounds good; When I say ambiguous I mean relative! (i.e. `1st jan`)

Comment (GuiltyDolphin) [inline t/DaysBetween.t]: @mintsoft Ah, alrighty. Though I've come up with some classifications for the Role:

- 'Formatted' - unambiguous, non--locale-based, and independent of time, eg., ISO8601.
- 'Relative' - may be locale-based, but is unambiguous for a given time. e.g., `tomorrow` or `in 3 days`.
- 'Descriptive' - may be locale-based, potentially ambiguous given the time e.g., `1st Jan` for which the year may vary.
- 'Locale' - 'formatted' dates that can vary from locale to locale.

Or at least... That's how I interpreted the existing set up and then progressed with development.

**Comment by NaveenKarippai** [comment]
@GuiltyDolphin could you please check the refactored test suite. Thanks!

File: t/DaysBetween.t [hunk 3]
```diff
@@ -10,143 +10,40 @@ use Test::MockTime qw( :all );
...
-            input     => ignore(),
-            operation => 'Days between',
-            result    => 15
-        },
-    ),
-    'days between jan 1 2012 and jan 1 123456' => undef,
-);
+sub build_structured_answer{
+    my($startDate, $endDate, $daysBetween, $inclusive) = @_;
+    return "There are $daysBetween days between $startDate and $endDate $inclusive.",
```
Comment (GuiltyDolphin) [inline t/DaysBetween.t]: We'll have trailing space here if there is no `$inclusive`.

**Comment by GuiltyDolphin** [comment]
@NaveenKarippai Apart from that, looks pretty good; I'll let @mintsoft sign off on this though.

File: lib/DDG/Goodie/DaysBetween.pm [hunk 4]
```diff
@@ -30,11 +30,15 @@ handle remainder => sub {
     my $startDate = date_output_string($date1);
     my $endDate   = date_output_string($date2);
 
-    return "There are $daysBetween days between $startDate and $endDate$inclusive.",
+    return "There are $daysBetween days between $startDate and $endDate $inclusive.",
```
Comment (mintsoft) [inline lib/DDG/Goodie/DaysBetween.pm]: As @GuiltyDolphin said this should really be:
`return "There are $daysBetween days between $startDate and $endDate$inclusive."` as it'll have trailing whitespace if it's not inclusive which'll look naff

File: t/DaysBetween.t [hunk 5]
```diff
@@ -10,143 +10,40 @@ use Test::MockTime qw( :all );
...
-            result    => 53
-        },
-    ),
+    'days between today and tomorrow'                           => build_test('14 Jul 2016', '15 Jul 2016', 1, ''),
+    'days between jan 1 and jan 15 inclusive'                   => build_test('01 Jan 2016', '15 Jan 2016', 15, $test_inclusive),
+    'days between jan 1 and 15th feb'                           => build_test('01 Jan 2016', '15 Feb 2016', 45, ''),
+    'how many days between feb 2 and feb 17'                    => build_test('02 Feb 2016', '17 Feb 2016', 15, ''),
+    'days between 2000-01-01 2001-01-01'                        => build_test('01 Jan 2000', '01 Jan 2001', 366, '' ),
+    'days between 2000-01-01 and 2001-01-01 inclusive'          => build_test('01 Jan 2000', '01 Jan 2001',367, $test_inclusive),
+    'daysbetween 2005-03-4 and 2020-11-8'                       => build_test('04 Mar 2005', '08 Nov 2020', 5728, ''),
```
Comment (mintsoft) [inline t/DaysBetween.t]: This should be `2005-03-04 and 2020-11-08`

File: t/DaysBetween.t [hunk 6]
```diff
@@ -10,143 +10,40 @@ use Test::MockTime qw( :all );
...
+    'days_between 2005-3-14 and 2003-1-2'                       => build_test('02 Jan 2003', '14 Mar 2005', 802, ''),
```
Comment (mintsoft) [inline t/DaysBetween.t]: Again `2005-03-14 and 2003-01-02`

File: t/DaysBetween.t [hunk 7]
```diff
@@ -10,143 +10,40 @@ use Test::MockTime qw( :all );
...
+    'days_between 2005-3-14 and 2003-1-2 inclusive'             => build_test('02 Jan 2003', '14 Mar 2005', 803, $test_inclusive),
```
Comment (mintsoft) [inline t/DaysBetween.t]: Same here `2005-03-14 and 2003-01-02`

File: t/DaysBetween.t [hunk 8]
```diff
@@ -10,143 +10,40 @@ use Test::MockTime qw( :all );
...
-        structured_answer => {
-            input     => ['22 May 2015', '14 Jul 2015'],
-            operation => 'Days between',
-            result    => 53
-        },
-    ),
+    'days between today and tomorrow'                           => build_test('14 Jul 2016', '15 Jul 2016', 1, ''),
+    'days between jan 1 and jan 15 inclusive'                   => build_test('01 Jan 2016', '15 Jan 2016', 15, $test_inclusive),
+    'days between jan 1 and 15th feb'                           => build_test('01 Jan 2016', '15 Feb 2016', 45, ''),
+    'how many days between feb 2 and feb 17'                    => build_test('02 Feb 2016', '17 Feb 2016', 15, ''),
```
Comment (mintsoft) [inline t/DaysBetween.t]: We've lost coverage here for queries in the format:
* `number of days from $date1 and $date2` 
* `number of days between $date1 and $date2`

File: t/DaysBetween.t (hunk 5 above)
Comment (GuiltyDolphin) [inline t/DaysBetween.t]: @mintsoft Is that not what it *already* is?

File: t/DaysBetween.t (hunk 8 above)
Comment (GuiltyDolphin) [inline t/DaysBetween.t]: Yah @NaveenKarippai make sure all the triggers are covered!

**Comment by NaveenKarippai** [comment]
@GuiltyDolphin @mintsoft all done! Please check the new changeset and Thanks for the review :smile:

**Comment by mintsoft** [comment]
@NaveenKarippai Almost there, but I noticed this:
extraspace

**Comment by NaveenKarippai** [comment]
@GuiltyDolphin fixed! Thanks again.

**Comment by mintsoft** [comment]
LGTM @GuiltyDolphin

**Comment by mintsoft** [comment]
Thanks @NaveenKarippai !

**Pull request merged**